src/dstruct/factor_graph/variable.h

```diff
@@ -54,10 +57,19 @@ namespace dd{
      * upper bound, initial value, current value, number of factors 
      */
     Variable(const long & _id, const int & _domain_type, 
-             const bool & _is_evid, const VariableValue & _lower_bound,
-             const VariableValue & _upper_bound, const VariableValue & _init_value, 
+             const bool & _is_evid, const int & cardinality,
+             const VariableValue & _init_value, 
              const VariableValue & _current_value, const int & _n_factors,
              bool is_observation);
+
+    // get the index-th value in the domain
+    inline VariableValue get_domain(int index) const {
+      return domain ? (*domain)[index] : index;
+    }
+    // get the index of the
+    inline int get_domain_index(int v) const {
+      return domain_map ? (*domain_map)[v] : (int)v;
```

Comment:
I thought treating pointers as boolean might generate warnings, but it seems people have mixed opinion: http://stackoverflow.com/questions/17772103/can-i-use-ifpointer-instead-of-ifpointer-null

Let's just keep it simple as it is then. :)